src/Filecoin/Address.cpp

```diff
@@ -0,0 +1,175 @@
+    }
+    if (!getType(data[0])) {
+        return false;
+    }
+    Type type = *getType(data[0]);
+    if (type == Type::ID) {
+        // Verify varuint encoding
+        if (data.size() > 11)
+            return false;
+        if (data.size() == 11 && data[10] > 0x01)
+            return false;
+        int i;
+        for (i = 1; i < data.size(); i++)
+            if ((data[i] & 0x80) == 0)
+                break;
+        return i == data.size() - 1;
+    } else {
+        return data.size() == 1 + Address::payloadSize(type);
+    }
+}
+
+static bool isValidID(const std::string& string) {
+    if (string.length() > 22)
+        return false;
+    for (int i = 2; i < string.length(); i++) {
+        if (string[i] < '0' || string[i] > '9') {
+            return false;
+        }
+    }
+    try {
+        size_t chars;
+        std::stoull(string.substr(2), &chars);
+        return chars > 0;
+    } catch (...) {
+        return false;
+    }
+}
+
+static bool isValidBase32(const std::string& string, Address::Type type) {
+    // Check if valid Base32.
+    uint8_t size = Address::payloadSize(type);
+    Data decoded;
+    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN)) {
+        return false;
+    }
+
+    // Check size
+    if (decoded.size() != size + 4) {
+        return false;
+    }
+
+    // Extract raw address.
+    Data address;
+    address.push_back(static_cast<uint8_t>(type));
+    address.insert(address.end(), decoded.data(), decoded.data() + size);
+
+    // Verify checksum.
+    Data should_sum = Hash::blake2b(address, checksumSize);
+    return std::memcmp(should_sum.data(), decoded.data() + size, checksumSize) == 0;
+}
+
+bool Address::isValid(const std::string& string) {
+    if (string.length() < 3) {
+        return false;
+    }
+    // Only main net addresses supported.
+    if (string[0] != PREFIX) {
+        return false;
+    }
+    // Get address type.
+    auto type = parseType(string[1]);
+    if (!type) {
+        return false;
+    }
+
+    // ID addresses are special, they are just numbers.
+    return type == Type::ID ? isValidID(string) : isValidBase32(string, *type);
+}
+
+Address::Address(const std::string& string) {
+    if (!isValid(string))
+        throw std::invalid_argument("Invalid address data");
+
+    Type type = *(parseType(string[1]));
+    // First byte is type
+    bytes.push_back(static_cast<uint8_t>(type));
+    if (type == Type::ID) {
+        uint64_t id = std::stoull(string.substr(2));
+        while (id >= 0x80) {
+            bytes.push_back(((uint8_t)id) | 0x80);
+            id >>= 7;
+        }
+        bytes.push_back((uint8_t)id);
+        return;
+    }
+
+    Data decoded;
+    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN))
+        throw std::invalid_argument("Invalid address data");
+    uint8_t payloadSize = Address::payloadSize(type);
+
+    bytes.insert(bytes.end(), decoded.data(), decoded.data() + payloadSize);
```

Comment:
Is a check needed for the case when decoded.size() < payloadSize?

Reply:
I don't think so since it is already checked in Address::isValidBase32